### ISSUE TITLE
[Windows] Windows keeps open a linked-to file

### ISSUE THREAD

**Issue by rly** [h5py/h5py#1569]
In the following code, I create a link from file2 to file1. When I then read file2 and access the link, even after I close file2, a process is still accessing file1. This seems to happen only on Windows and not MacOS or Linux. 

```python
import h5py
import os

with h5py.File('file1.h5', 'w') as f1:
    grp = f1.create_group('group')

with h5py.File('file2.h5', 'w') as f2:
    f2['link'] = h5py.ExternalLink("file1.h5", "/group")

with h5py.File('file2.h5', 'r') as f2:
    obj = f2['link']

print(obj)
# obj.file.close()  # this or the line below is required to release the linked-to file only on Windows
# del(obj)

os.remove('file1.h5')
# f3 = h5py.File('file1.h5', 'w')
```

On `os.remove`:
```
<HDF5 group "/group" (0 members)>
Traceback (most recent call last):
  File "h5py_file_test.py", line 16, in <module>
    os.remove('file1.h5')
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'file1.h5'
```

If I comment out `os.remove` and uncomment the next line to open file1 in 'w' mode:
```
Traceback (most recent call last):
  File "h5py_file_test.py", line 18, in <module>
    f3 = h5py.File('file1.h5', 'w')
  File "C:\Users\Ryan\Miniconda3\envs\dev\lib\site-packages\h5py\_hl\files.py", line 408, in __init__
    swmr=swmr)
  File "C:\Users\Ryan\Miniconda3\envs\dev\lib\site-packages\h5py\_hl\files.py", line 179, in make_fid
    fid = h5f.create(name, h5f.ACC_TRUNC, fapl=fapl, fcpl=fcpl)
  File "h5py\_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py\_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py\h5f.pyx", line 108, in h5py.h5f.create
OSError: Unable to create file (unable to truncate a file which is already open)
```

So it seems like on Windows, h5py keeps the linked-to file open in read mode in the current Python process. Using `obj.file.close()`, `del(obj)`, or ending the Python process resolves the issue. This is fine on MacOS and Linux.

This might be related to https://github.com/h5py/h5py/issues/1012

 * Operating System: Windows 10
 * Python version: 3.7.7
 * Where Python was acquired: Miniconda for Windows
 * h5py version: 2.10.0
 * HDF5 version: 1.10.5

**Comment by vasole** [comment]
If after print(obj) you make obj = None, you will be able to delete the file.

You are keeping a reference to an object inside a file. I am not surprised the file cannot be deleted.

**Comment by rly** [comment]
My point is that the behavior is inconsistent across OS. On MacOS/Linux, if a reference exists to an object inside the file, that file can still be deleted. Perhaps that is an issue with how different OS's handle open files and cannot be worked around.

**Comment by takluyver** [comment]
Yes, that's a difference in how operating systems handle files. It's not specific to HDF5: Windows prevents deleting a file that's open, Linux & Mac don't. I don't think it makes sense to try and explain this in the h5py docs.

What may be worth explaining is when h5py & HDF5 keep a file open - see also #1500.

**Comment by takluyver** [comment]
PR #1570 closed #1500, so I'm going to close this one as well.
